core/src/main/scala/kafka/tools/GetOffsetShell.scala

```diff
@@ -132,23 +161,85 @@ object GetOffsetShell {
-    if (partitionInfos.isEmpty)
-      None
-    else if (partitionIds.isEmpty)
-      Some(partitionInfos)
-    else
-      Some(partitionInfos.filter(p => partitionIds.contains(p.partition)))
+  def createTopicPartitionFilterWithPatternList(topicPartitions: String, excludeInternalTopics: Boolean): PartitionInfo => Boolean = {
+    val ruleSpecs = topicPartitions.split(",")
+    val rules = ruleSpecs.map { ruleSpec => parseRuleSpec(ruleSpec, excludeInternalTopics) }
+    tp => rules.exists(rule => rule.apply(tp))
   }
 
+  def parseRuleSpec(ruleSpec: String, excludeInternalTopics: Boolean): PartitionInfo => Boolean = {
+    def wrapNullOrEmpty(s: String): Option[String] = {
+      if (s == null || s.isEmpty)
+        None
+      else
+        Some(s)
+    }
+
+    val matcher = topicPartitionPattern.matcher(ruleSpec)
+    if (!matcher.matches() || matcher.groupCount() == 0)
+      throw new IllegalArgumentException(s"Invalid rule specification: $ruleSpec")
+
+    val topicPattern = wrapNullOrEmpty(matcher.group(1))
+    val lowerRange = if (matcher.groupCount() >= 2) wrapNullOrEmpty(matcher.group(2)) else None
+    val upperRangeSection = if (matcher.groupCount() >= 3) matcher.group(3) else null
+    val upperRange = wrapNullOrEmpty(upperRangeSection)
+    val isRange = upperRangeSection != null
+
+    val includeList = IncludeList(topicPattern.getOrElse(".*"))
+
+    val partitionFilter: Int => Boolean = if (lowerRange.isEmpty && upperRange.isEmpty) {
+      _ => true
+    } else if (lowerRange.isEmpty) {
+      val upperBound = upperRange.get.toInt
+      p => p < upperBound
+    } else if (upperRange.isEmpty) {
+      val lowerBound = lowerRange.get.toInt
+      if (isRange) {
+        p => p >= lowerBound
+      }
+      else {
+        p => p == lowerBound
+      }
+    } else {
+      val upperBound = upperRange.get.toInt
+      val lowerBound = lowerRange.get.toInt
+      p => p >= lowerBound && p < upperBound
+    }
+
+    tp => includeList.isTopicAllowed(tp.topic(), excludeInternalTopics) && partitionFilter.apply(tp.partition())
+  }
+
+  /**
+   * Creates a topic-partition filter based on a topic pattern and a set of partition ids.
+   */
+  def createTopicPartitionFilterWithTopicAndPartitionPattern(topicOpt: Option[String], excludeInternalTopics: Boolean, partitionIds: Set[Int]): PartitionInfo => Boolean = {
+    val topicsFilter = IncludeList(topicOpt.getOrElse(".*"))
+    t => topicsFilter.isTopicAllowed(t.topic, excludeInternalTopics) && (partitionIds.isEmpty || partitionIds.contains(t.partition))
+  }
+
+  /**
+   * Return the partition infos. Filter them with topicPartitionFilter.
+   */
+  private def listPartitionInfos(consumer: KafkaConsumer[_, _], topicPartitionFilter: PartitionInfo => Boolean): Seq[PartitionInfo] = {
+    consumer.listTopics.asScala.values.flatMap { partitions =>
+      partitions.asScala.filter { tp => topicPartitionFilter.apply(tp) }
```

Comment:
nit: We should use parenthesis here instead of curly braces. `filter(topicPartitionFilter)` might even work directly as well.